### PR TITLE
Support vectorized interpolation with more scipy interpolators

### LINKED ISSUE #9049: Cannot use documented interp() methods due to "vectorizeable_only" check and kwargs name clash
### What happened?

I would like to use `xarray.Dataset.interp()` to interpolate data using the scipy "makima" interpolator, `scipy.interpolate.Akima1DInterpolator(... method='makima')`. It seems to me this is possible according to the documentation here:
https://docs.xarray.dev/en/latest/generated/xarray.Dataset.interp.html
https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.Akima1DInterpolator.html#scipy.interpolate.Akima1DInterpolator

However, I am not able to do this, I think due to two separate problems:
1) There is a check that only allows "vectorizeable interpolators" for one-dimensional interpolation, the motivation for which I do not understand. This makes 'akima', 'spline', and other documented interp methods unavailable. 

2) The way kwargs are passed to the scipy interpolator results in multiple kwargs named "method", making specifically "makima" (modified akima) unavailable. 

### What did you expect to happen?

I expect interpolation to happen. 

### Minimal Complete Verifiable Example

```Python
## create a Dataset and new x array for interpolation
import numpy as np
import xarray as xr
x = np.arange(0, 5, 0.05)
x_dense = np.arange(0, 5, 0.01)
y = np.sin(x)
ds = xr.Dataset({'y': (('x'), y)}, coords={'x': x})

## do a cubic interpolation. this works as expected
ds_interp_cubic = ds.interp(x=x_dense, method='cubic')

## attempt akima interpolation. this fails.
ds_interp_akima = ds.interp(x=x_dense, method='akima')

## attempt makima interpolation. this also fails.
ds_interp_makima = ds.interp(x=x_dense, method='akima', kwargs={'method': 'makima'})
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

```Python
## For the akima failure, this is the ipython traceback:

File /opt/homebrew/Caskroom/miniforge/base/envs/npaccel-larch/lib/python3.11/site-packages/xarray/core/dataset.py:4003, in Dataset.interp(self, coords, method, assume_sorted, kwargs, method_non_numeric, **coords_kwargs)
   4000 if dtype_kind in "uifc":
   4001     # For normal number types do the interpolation:
   4002     var_indexers = {k: v for k, v in use_indexers.items() if k in var.dims}
-> 4003     variables[name] = missing.interp(var, var_indexers, method, **kwargs)
   4004 elif dtype_kind in "ObU" and (use_indexers.keys() & var.dims):
   4005     # For types that we do not understand do stepwise
   4006     # interpolation to avoid modifying the elements.
   (...)
   4009     # this loop there might be some duplicate code that slows it
   4010     # down, therefore collect these signals and run it later:
   4011     reindex = True

File /opt/homebrew/Caskroom/miniforge/base/envs/npaccel-larch/lib/python3.11/site-packages/xarray/core/missing.py:632, in interp(var, indexes_coords, method, **kwargs)
    630 original_dims = broadcast_dims + dims
    631 new_dims = broadcast_dims + list(destination[0].dims)
--> 632 interped = interp_func(
    633     var.transpose(*original_dims).data, x, destination, method, kwargs
    634 )
    636 result = Variable(new_dims, interped, attrs=var.attrs, fastpath=True)
    638 # dimension of the output array

File /opt/homebrew/Caskroom/miniforge/base/envs/npaccel-larch/lib/python3.11/site-packages/xarray/core/missing.py:687, in interp_func(var, x, new_x, method, kwargs)
    684     return var.copy()
    686 if len(x) == 1:
--> 687     func, kwargs = _get_interpolator(method, vectorizeable_only=True, **kwargs)
    688 else:
    689     func, kwargs = _get_interpolator_nd(method, **kwargs)

File /opt/homebrew/Caskroom/miniforge/base/envs/npaccel-larch/lib/python3.11/site-packages/xarray/core/missing.py:494, in _get_interpolator(method, vectorizeable_only, **kwargs)
    492     interp_class = ScipyInterpolator
    493 elif vectorizeable_only:
--> 494     raise ValueError(
    495         f"{method} is not a vectorizeable interpolator. "
    496         f"Available methods are {interp1d_methods}"
    497     )
    498 elif method == "barycentric":
    499     interp_class = _import_interpolant("BarycentricInterpolator", method)

ValueError: akima is not a vectorizeable interpolator. Available methods are ('linear', 'nearest', 'zero', 'slinear', 'quadratic', 'cubic', 'polynomial')


## for the makima failure, this is the ipython traceback:

/opt/homebrew/Caskroom/miniforge/base/envs/npaccel-larch/lib/python3.11/site-packages/xarray/core/dataset.py in ?(self, coords, method, assume
_sorted, kwargs, method_non_numeric, **coords_kwargs)
   3999             dtype_kind = var.dtype.kind
   4000             if dtype_kind in "uifc":
   4001                 # For normal number types do the interpolation:
   4002                 var_indexers = {k: v for k, v in use_indexers.items() if k in var.dims}
-> 4003                 variables[name] = missing.interp(var, var_indexers, method, **kwargs)
   4004             elif dtype_kind in "ObU" and (use_indexers.keys() & var.dims):
   4005                 # For types that we do not understand do stepwise
   4006                 # interpolation to avoid modifying the elements.

TypeError: interp() got multiple values for argument 'method'
```

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:37:07) [Clang 15.0.7 ]
python-bits: 64
OS: Darwin
OS-release: 23.5.0
machine: arm64
processor: arm
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.3
libnetcdf: None

xarray: 2024.3.0
pandas: 2.2.2
numpy: 1.26.4
scipy: 1.13.0
netCDF4: None
pydap: None
h5netcdf: None
h5py: 3.11.0
Nio: None
zarr: None
cftime: None
nc_time_axis: None
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: 3.8.4
cartopy: None
seaborn: None
numbagg: None
fsspec: 2024.3.1
cupy: None
pint: 0.23
sparse: None
flox: 0.9.6
numpy_groupies: 0.11.1
setuptools: 69.5.1
pip: 24.0
conda: None
pytest: None
mypy: None
IPython: 8.22.2
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by hollymandel**
- [x] Closes #9049
- [ ] Adds support for vectorized (broadcast) 1d interpolation using SciPy interpolants, including "akima" and "makima"
- [ ] Does not add support for deprecated SplineInterpolator (but upholds preexisting functionality). I wanted to add a deprecation warning but this caused tests to fail so I left it commented.
- [ ] Akima and makima cases added to test_interp.py::test_interpolate_vectorize

### CODE REVIEW

**Comment by max-sixty** [comment]
Something has gone kinda wrong with git here! Maybe try squashing to a single commit?

**Comment by hollymandel** [comment]
> Something has gone kinda wrong with git here! Maybe try squashing to a single commit?

yes, done, thanks!

**Comment by hollymandel** [comment]
Sounds good! I actually wanted it to be a "draft PR" as I'm still double-checking it, though perhaps submitting it this way is confusing on the maintainers' end and I should rescind until it's ready.

**Comment by keewis** [comment]
> though perhaps submitting it this way is confusing on the maintainers' end and I should rescind until it's ready

we're actually happy to have PRs developed like this, it allows us to guide people more easily in case there's any issues. This includes advising in case the direction is seriously wrong (which I don't think it is in this case, but I also don't know this area too well)

File: xarray/core/missing.py [hunk 1]
```diff
@@ -203,6 +203,7 @@ def __init__(
 
         self.method = method
         self.cons_kwargs = kwargs
+        del self.cons_kwargs["axis"]
```
Comment (dcherian) [inline xarray/core/missing.py:206]: why is "axis" being sent to this level if it isn't needed? This suggests we need to delete it elsewhere

File: xarray/core/missing.py [hunk 2]
```diff
@@ -489,25 +491,31 @@ def _get_interpolator(
...
             interp_class = _import_interpolant("PchipInterpolator", method)
         elif method == "spline":
+            # utils.emit_user_level_warning(
+            #     "The 1d SplineInterpolator class is performing an incorrect calculation and "
+            #     "is being deprecated. Please use `method=polynomial` for 1D Spline Interpolation.",
+            #     PendingDeprecationWarning,
+            # )
+            if vectorizeable_only:
+                raise ValueError(f"{method} is not a vectorizeable interpolator. ")
             kwargs.update(method=method)
```
Comment (dcherian) [inline xarray/core/missing.py:512]:
```suggestion
            kwargs.update(method=method)
            kwargs.pop("axis")
```

Comment (dcherian) [inline xarray/core/missing.py:512]:
```suggestion
            kwargs.update(method=method)
            kwargs.pop("axis")
```

File: xarray/core/missing.py [hunk 3]
```diff
@@ -614,9 +623,6 @@ def interp(var, indexes_coords, method: InterpOptions, **kwargs):
     if not indexes_coords:
         return var.copy()
 
-    # default behavior
```
Comment (dcherian) [inline xarray/core/missing.py:617]: nice cleanup!

File: xarray/tests/test_interp.py [hunk 4]
```diff
@@ -132,29 +132,55 @@ def func(obj, new_x):
     assert_allclose(actual, expected)
 
 
-@pytest.mark.parametrize("use_dask", [False, True])
-def test_interpolate_vectorize(use_dask: bool) -> None:
+@pytest.mark.parametrize(
```
Comment (dcherian) [inline xarray/tests/test_interp.py:135]:
```suggestion
@requires_scipy
@pytest.mark.parametrize(
```

File: xarray/tests/test_interp.py [hunk 5]
```diff
@@ -132,29 +132,55 @@ def func(obj, new_x):
...
+        (False, "linear"),
+        (False, "akima"),
+        (False, "makima"),
+        (True, "linear"),
+        (True, "akima"),
+    ),
+)
+def test_interpolate_vectorize(use_dask: bool, method: str) -> None:
     if not has_scipy:
         pytest.skip("scipy is not installed.")
```
Comment (dcherian) [inline xarray/tests/test_interp.py:147]:
```suggestion
```

driveby cleanup. The `requires_scipy`decorator will skip if scipy isn't available. Import it  from `xarray.tests` if it's not present.

File: xarray/tests/test_interp.py [hunk 6]
```diff
@@ -132,29 +133,60 @@ def func(obj, new_x):
...
 
-    if not has_dask and use_dask:
-        pytest.skip("dask is not installed in the environment.")
-
     # scipy interpolation for the reference
-    def func(obj, dim, new_x):
+    def func(obj, dim, new_x, method):
+        scipy_kwargs = {}
+        interpolant_options = {
+            "barycentric": "BarycentricInterpolator",
```
Comment (dcherian) [inline xarray/tests/test_interp.py:162]: nit: slightly cleaner to do this instead:
```suggestion
        interpolators = {
            "barycentric": scipy.interpolate.BarycentricInterpolator,
```

File: xarray/tests/test_interp.py [hunk 7]
```diff
@@ -132,29 +133,60 @@ def func(obj, new_x):
...
+            return interpolant(
+                da[dim], obj.data, axis=obj.get_axis_num(dim), **scipy_kwargs
+            )(new_x).reshape(shape)
+        else:
+            scipy_kwargs["kind"] = method
+            scipy_kwargs["bounds_error"] = False
+            scipy_kwargs["fill_value"] = np.nan
+            return scipy.interpolate.interp1d(
+                da[dim], obj.data, axis=obj.get_axis_num(dim), **scipy_kwargs
+            )(new_x).reshape(shape)
```
Comment (dcherian) [inline xarray/tests/test_interp.py:186]: another nit: this reduces a little bit of indirection and makes it easier to review

```suggestion
            return scipy.interpolate.interp1d(
                da[dim], obj.data, axis=obj.get_axis_num(dim), kind=method, bounds_error=False, fill_value=np.nan
            )(new_x).reshape(shape)
```

File: xarray/core/missing.py [hunk 8]
```diff
@@ -755,8 +761,9 @@ def interp_func(var, x, new_x, method: InterpOptions, kwargs):
 
 def _interp1d(var, x, new_x, func, kwargs):
     # x, new_x are tuples of size 1.
-    x, new_x = x[0], new_x[0]
-    rslt = func(x, var, assume_sorted=True, **kwargs)(np.ravel(new_x))
+    x, new_x = x[0].data, new_x[0].data
```
Comment (dcherian) [inline xarray/core/missing.py:764]: Is there a reason you got rid of `assume_sorted=True`

File: xarray/core/missing.py (hunk 2 above)
Comment (dcherian) [reply xarray/core/missing.py:512]: this will fix your tests

File: xarray/core/missing.py (hunk 8 above)
Comment (hollymandel) [reply xarray/core/missing.py:764]: 1. This doesn't seem to have been necessary, changing back

2. The interpolators I have added (`Scipy.Interpolate.Akima1DInterpolator`, etc...) require sorted axes but don't take `assume_sorted` as an argument. Only `Scipy.Interpolate.Interp1d` takes this argument. Since we are always calling it with `True`,  I'm just letting the default for ScipyInterpolator init impose this. 

There is maybe a philosophical conflict here and in previous comments about kwarg handling for the different interpolators. The preexisting script (`missing.py`) has special classes (`NumpyInterpolator`, `ScipyInterpolator`...) which basically just do this handling. I didn't add any new classes for the new interpolators and rather tried to do everything with kwargs determined by `_get_interpolator`.

Comment (dcherian) [reply xarray/core/missing.py:764]: Re (1): If you're comfortable with typing, some type hints would go a long way towards improving this module. 

Re (2): sounds good

Comment (hollymandel) [reply xarray/core/missing.py]: (1) Yeah can do! Prefer a separate commit though (?)

**Comment by dcherian** [comment]
(fyi we squash on merge, so you needn't squash every change)

File: xarray/core/missing.py (hunk 8 above)
Comment (dcherian) [reply xarray/core/missing.py]: we squash merge so it doesn't matter :) unless you meant separate PR, in which case that would be fine too.

**Review by dcherian — APPROVED** [review]
Very nice work! Thanks for taking this on, we've needed someone to do this for a while now :)

One last request, please add a note to `whats-new.rst`

**Comment by max-sixty** [comment]
Thanks @hollymandel !
